runtime/gc_vlhgc/SchedulingDelegate.cpp

```diff
@@ -1017,21 +1294,229 @@ MM_SchedulingDelegate::calculateEdenSize(MM_EnvironmentVLHGC *env)
 	Trc_MM_SchedulingDelegate_calculateEdenSize_Exit(env->getLanguageVMThread(), (_edenRegionCount * regionSize));
 }
 
-UDATA
+void
+MM_SchedulingDelegate::moveTowardRecommendedEden(MM_EnvironmentVLHGC *env, double edenChangeSpeed) 
+{
+	Assert_MM_true(edenChangeSpeed <= 1 && edenChangeSpeed >= 0);
+
+	if ((0 == _historicalPartialGCTime) || (0 == _averagePgcInterval)) {
+		/* Until we have collected any information about PGC time, we don't have the data we need to make informed decision about eden size  */
+		return;
+	}
+
+	uintptr_t recommendedEdenSizeBytes = calculateRecommendedEdenSize(env);
+
+	uintptr_t currentIdealEdenBytes = getIdealEdenSizeInBytes(env);
+	uintptr_t currentIdealEdenRegions = _idealEdenRegionCount;
+
+	/* 
+	 * The closer edenChangeSpeed is to 1, the larger the move towards recommendedEdenSizeBytes will be. 
+	 * 1 implies that eden should move all the way towards recommendedEdenSizeBytes.
+	 */
+	intptr_t edenChange = (intptr_t)recommendedEdenSizeBytes - currentIdealEdenBytes;
+	intptr_t targetEdenChange = (intptr_t)(edenChange * edenChangeSpeed);
+	uintptr_t targetEdenBytes = currentIdealEdenBytes + targetEdenChange;
+	uintptr_t targetEdenRegions = targetEdenBytes / _regionManager->getRegionSize();
+
+	_edenSizeFactor = (intptr_t)targetEdenRegions - currentIdealEdenRegions;
```

Comment:
For more consistent code at call site (checkEdenSizeAfterPgc), let's return the value and make the caller store it in _edenSizeFactor. The caller already deals with this factor in non-fully-expanded case.

Reply:
100% agreed. Will make that change